R/calculate_tdm.R

```diff
@@ -0,0 +1,305 @@
+#'   in addition to these ones (which are always used):
+#'
+#'   `r toString(crucial_tdm_groups())`
+#'
+#' @return A tibble with the columns specified in the `...` input, as well as:
+#'   `tdm_tech`, the technology level transition disruption metric and
+#'   `tdm_sec`, the sector level transition disruption metric.
+#' @export
+#'
+#' @examples
+#' library(tibble)
+#'
+#' pacta_results <- tibble(
+#'   investor_name = rep("some_investor", 3),
+#'   portfolio_name = rep("some_portfolio", 3),
+#'   allocation = rep("portfolio_weight", 3),
+#'   equity_market = rep("GlobalMarket", 3),
+#'   scenario_geography = rep("Global", 3),
+#'   ald_sector = rep("Power", 3),
+#'   technology = rep("RenewablesCap", 3),
+#'   year = c(2020, 2025, 2030),
+#'   plan_alloc_wt_tech_prod = rep(1, 3),
+#'   scen_alloc_wt_tech_prod = 1:3,
+#'   plan_carsten = rep(0.5, 3)
+#' )
+#' pacta_results
+#'
+#' calculate_tdm(pacta_results, t0 = 2020)
+calculate_tdm <- function(data, t0, t1 = 5, t2 = 10, additional_groups = NULL) {
+  stopifnot(
+    is.data.frame(data),
+    is.numeric(t0),
+    is.numeric(t1),
+    is.numeric(t2)
+  )
+
+  if (!is.null(additional_groups)) stopifnot(is.character(additional_groups))
+
+  groups <- union(crucial_tdm_groups(), additional_groups)
+
+  check_calculate_tdm(data, t0, t1, t2, groups)
+
+  filtered_data <- filter(data, .data$allocation == "portfolio_weight")
+
+  if (nrow(filtered_data) == 0) {
+    return(warn_zero_rows(tdm_prototype()))
+  }
+
+  # TODO: Filter the data fora  particular scenario. Function should gain the
+  # argument `scenario`, which by default specifies IPR - FPS scenario. I am
+  # waiting until the data is prepared as I'm not sure exactly what this
+  # scenario will be called in the input data.
+
+  data_with_monotonic_factors <- add_monotonic_factor(filtered_data, t0, t1, t2, groups)
+
+  initial_year_data <- filter(data_with_monotonic_factors, .data$year == t0)
+
+  preformatted_data <- pre_format_data(data_with_monotonic_factors, t0, t1, t2, groups)
+
+  data_with_tdm <- add_tdm(preformatted_data, groups)
+
+  formatted_data_with_tdm <- left_join(
+    initial_year_data,
+    data_with_tdm,
+    by = groups
+  )
+
+  formatted_data_with_tdm %>%
+    add_aggregate_tdm(t0, groups) %>%
+    select(names(tdm_prototype()), all_of(groups))
+}
+
+warn_zero_rows <- function(data) {
+  # NOTE: This function will only work if the allocation method is
+  # portfolio_weight. ownership_weight outputs 0 for all carsten metric values,
+  # and thus we would be dividing by zero otherwise...
+  message <- 'Filtering for "portfolio_weight" allocation, outputs 0 rows'
+  warn(message, class = "has_zero_rows")
+
+  # Pass `data` to inline inside return()
+  invisible(data)
+}
+
+add_time_step <- function(data, t0, t1, t2) {
+  data %>%
+    mutate(
+      time_step = case_when(
+        .data$year == t0 ~ "t0",
+        .data$year == t0 + t1 ~ "plus_t1",
+        .data$year == t0 + t2 ~ "plus_t2"
```

Comment:
isn't it more important to use the `.env$` pronoun than the `.data$` since the `.data$` variables are used by default?

Reply:
afaik this is best practice, but i'm gonna divert that question to @maurolepore for a better explanation

Reply:
I think the `.data$` usage is "best practice" because it avoids `R CMD check` warnings.

I personally believe it's "good practice" to call environment variables (versus data variables) explicitly using the `.env$` pronoun because by default dplyr will use the data variable if there is a data variable and an environment variable with the same name.

e.g.
```r
library(dplyr)

t0 <- 2020

tibble(x = "ABC", t0 = 1984) %>% 
  mutate(t1 = t0 + 5)
# # A tibble: 1 x 3
#   x        t0    t1
#   <chr> <dbl> <dbl>
# 1 ABC    1984  1989

tibble(x = "ABC", t0 = 1984) %>% 
  mutate(t1 = .env$t0 + 5)
# # A tibble: 1 x 3
#   x        t0    t1
#   <chr> <dbl> <dbl>
# 1 ABC    1984  2025
```

Reply:
ah ok, i mean in that case we definitely want `.data` here and not `.env`.

Reply:
so in the case of `.data$year == t0 ~ "t0",` you want `t0` to be the column in the data, not the `t0` that is passed as an argument?